src/main/scala/com/databricks/spark/avro/avroRelation2/AvroRelation2.scala

```diff
@@ -0,0 +1,216 @@
+    val outputAvroSchema = SchemaConverters.convertStructToAvro(dataSchema, build, recordNamespace)
+    AvroJob.setOutputKeySchema(job, outputAvroSchema)
+    val AVRO_COMPRESSION_CODEC = "spark.sql.avro.compression.codec"
+    val AVRO_DEFLATE_LEVEL = "spark.sql.avro.deflate.level"
+    val COMPRESS_KEY = "mapred.output.compress"
+
+    sqlContext.getConf(AVRO_COMPRESSION_CODEC, "snappy") match {
+      case "uncompressed" =>
+        logInfo("writing Avro out uncompressed")
+        job.getConfiguration.setBoolean(COMPRESS_KEY, false)
+      case "snappy" =>
+        logInfo("using snappy for Avro output")
+        job.getConfiguration.setBoolean(COMPRESS_KEY, true)
+        job.getConfiguration.set(AvroJob.CONF_OUTPUT_CODEC, DataFileConstants.SNAPPY_CODEC)
+      case "deflate" =>
+        val deflateLevel = sqlContext.getConf(
+          AVRO_DEFLATE_LEVEL, Deflater.DEFAULT_COMPRESSION.toString).toInt
+        logInfo(s"using deflate: $deflateLevel for Avro output")
+        job.getConfiguration.setBoolean(COMPRESS_KEY, true)
+        job.getConfiguration.set(AvroJob.CONF_OUTPUT_CODEC, DataFileConstants.DEFLATE_CODEC)
+        job.getConfiguration.setInt(AvroOutputFormat.DEFLATE_LEVEL_KEY, deflateLevel)
+      case unknown: String => sys.error(s"Unknown output compression: $unknown")
+    }
+    new AvroOutputWriterFactory(dataSchema, recordName, recordNamespace)
+  }
+
+
+  /**
+   * This filters out unneeded columns before converting into the internal row representation.
+   * The first record is used to get the sub-schema that contains only the requested fields,
+   * this is then used to generate the field converters and the rows that only
+   * contain `requiredColumns`
+   */
+  override def buildScan(requiredColumns: Array[String], inputs: Array[FileStatus]): RDD[Row] = {
+    if (inputs.isEmpty) {
+      sqlContext.sparkContext.emptyRDD[Row]
+    } else {
+      inputs.map(path =>
+        sqlContext.sparkContext.hadoopFile(
+          path.getPath.toString,
+          classOf[org.apache.avro.mapred.AvroInputFormat[GenericRecord]],
+          classOf[org.apache.avro.mapred.AvroWrapper[GenericRecord]],
+          classOf[org.apache.hadoop.io.NullWritable]).keys.map(_.datum())
+          .mapPartitions { records =>
+            if (records.isEmpty) {
+              Iterator.empty
+            } else {
+              val first = records.next()
+              val superSchema = first.getSchema // the schema of the actual record
+              // the fields that are actually required along with their converters
+              val avroFields = superSchema.getFields
+              val fields = requiredColumns.map { column =>
+                avroFields.collectFirst {
+                  case f if f.name == column =>
+                    val newField = new Field(f.name, f.schema, f.doc,f.defaultValue, f.order)
+                    (SchemaConverters.createConverterToSQL(newField.schema), newField)
+                }.get // required to be there
```

Comment:
This will not fail since `requiredColumns` is an empty list and the `.get` will never get called. I have added a test for it just in case, which I will submit shortly.